hashintel/hash
# Merge Search components for better DOM reuse.

## 🌟 What is the purpose of this PR?




@benwerner01 pointed out in a comment on the original [Header PR](https://github.com/hashintel/hash/pull/358#discussion_r817763041) that the mobile and desktop versions of search should be merged to promote component reuse. The merge should also consider having both versions of search on the same DOM level so that React reuses the same DOM node. This has benefits in maintaining browser events like input focus on the same element across screen size shifts.

So, this PR tries to consolidate both versions of the search input into a single component which maintains DOM hierarchy, and hence, is reused. Changes made in this PR:
- [If a user is in mobile mode and has not clicked the search icon](https://github.com/hashintel/hash/pull/375/files#diff-f6666bfe47541e94befe2f523dbbc125b1e02c578a1357d8e56a70b0d45a1bc5R181), display the search icon instead of the search input.
- [Display the "cancel" button](https://github.com/hashintel/hash/pull/375/files#diff-f6666bfe47541e94befe2f523dbbc125b1e02c578a1357d8e56a70b0d45a1bc5R208) conditionally for mobile.
- [Change input width conditionally](https://github.com/hashintel/hash/pull/375/files#diff-ebccff6a5e8b22ec9fa2bd595f473d4b51efdf431feaf92d85f3fc3fd6adcccbR25) here and on `InputBase`.
- [Display `endAdornment`](https://github.com/hashintel/hash/pull/375/files#diff-ebccff6a5e8b22ec9fa2bd595f473d4b51efdf431feaf92d85f3fc3fd6adcccbR68) on desktop.
## Blocked By

- [x] - https://github.com/hashintel/hash/pull/378

## ❓ How to test this?



1. Checkout the branch locally.
2. Click on the search input and start typing.
3. Try to change screen size without losing focus on the input.
4. Input focus should be preserved across desktop and mobile.

## 📹 Demo



> Note: In the demo below, wait for a few seconds to see how the input focus is preserved across screen sizes.

![chrome_AAjXTSVoF4](https://user-images.githubusercontent.com/22196279/157040844-9b94f8fa-9918-4464-84c7-0f0f228f9041.gif)



---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201925844970634
